operator/src/main/java/oracle/kubernetes/operator/steps/ManagedServerUpIteratorStep.java

```diff
@@ -142,19 +154,53 @@ private boolean isServerInCluster(ServerStartupInfo ssi) {
 
   static class StartManagedServersStep extends Step {
     final Collection<StepAndPacket> startDetails;
+    final Queue<StepAndPacket> startDetailsQueue = new ConcurrentLinkedQueue<>();
+    final String clusterName;
+    int numStarted = 0;
+    int maxConcurrency = 0;
 
-    StartManagedServersStep(Collection<StepAndPacket> startDetails, Step next) {
+    StartManagedServersStep(String clusterName, Collection<StepAndPacket> startDetails, Step next) {
       super(next);
+      this.clusterName = clusterName;
       this.startDetails = startDetails;
+      startDetails.forEach(this::add);
     }
 
-    Collection<StepAndPacket> getStartDetails() {
-      return startDetails;
+    void add(StepAndPacket serverToStart) {
+      startDetailsQueue.add(new StepAndPacket(serverToStart.step, serverToStart.packet));
+      this.maxConcurrency = Optional.ofNullable(
+              (Integer) serverToStart.packet.get(ProcessingConstants.MAX_CONCURRENCY)).orElse(0);
     }
 
     @Override
     public NextAction apply(Packet packet) {
-      return doForkJoin(new ManagedServerUpAfterStep(getNext()), packet, startDetails);
+
+      if (startDetailsQueue.isEmpty()) {
+        return doNext(new ManagedServerUpAfterStep(getNext()), packet);
+      } else if (isServiceOnlyOrShuttingDown()) {
+        Collection<StepAndPacket> servers = Collections.singletonList(startDetailsQueue.poll());
+        return doForkJoin(this, packet, servers);
+      } else if (serverAvailableToStart(packet.getSpi(DomainPresenceInfo.class))) {
+        this.numStarted++;
```

Comment:
What is the concurrency of this code? Will multiple threads ever be running here?

Reply:
It's one thread per cluster and one thread for non-clustered servers. It'll be single threaded for one cluster and 2 threads for 2 cluster case. If domain also have non-clustered (standalone) servers, it'll create a separate fiber for those servers.

Reply:
Let me be clearer in my question. Do I need to worry about the concurrency of updating the numStarted variable? I can't really tell from your answer.

Reply:
Sorry for not being clear in my previous answer. In 2 cluster scenario, 2 threads could reading/updating in parallel but overall concurrency will be low. I have `numStarted` to AtomicInteger as we discussed and made `numConcurrency ` final by setting it in constructor. Please let me know if I missed anything. Thanks.